juju-tutorials/juju-tricks.md

```diff
@@ -0,0 +1,177 @@
+[comment]: <> (Juju Tricks)
+
+## Introduction
+
+Duration: 0:01
+
+This article is compilation of CLI tricks, useful for recovery of juju deployments after OS services' hosts reboot / crash / network splits / etc.
+
+### Requirements
+
+There is no specific setup precedure for this tutorial because it contains commands that are applicable in various situations in any juju deployment. Most of the commands are using just standard cli utilities like `grep`, `sed` or `awk`.
+
+[note status="What to do if you get stuck"]
+The best place to ask for help is the [Charmhub Discourse forum](https://discourse.charmhub.io/).
+
+If you prefer chatting, visit us on [IRC](https://webchat.freenode.net/#juju).
+[/note]
+
+## Tricks on broken units / machines / models
+
+Duration: N/A
+
+[note status="Direct reachability"]
+For the commands in this section to work, the targeted units or machines must be directly reachable via `juju ssh`. If, for example, LXCs are using network that is reachable only from the LXC host machine, you will encounter connectivity errors.
+[/note]
+
+### Restarting units and machines based on their state
+
+To restart units that are in particular state (e.g. `lost` or `down`) use following command and adjust the first part, `state=<TARGET_STATE>`, to match state of the units that you want to restart. The command, as it is, will restart units in state `down`.
+
+```console
+state=down; juju status --format oneline | grep "agent:${state}" | xargs -l4 | column -t | egrep -o '[a-z0-9-]+/[0-9]+' | sort -u | xargs -P4 -I@ bash -c 'unit=@; juju ssh $unit "sudo service jujud-unit-${unit/\//-} restart"'
+```
+
+To restart machine in particular state (including LXCs) use the following command and adjust the first part, `state=<TARGET_STATE>`, to match state of the machines that you want to restart. The command, as it is, will restart machines in state `down`.
+
+```console
+state=down; juju machines | grep -v 'Machine' | awk -v state=${state} '{if ($2 == state) print $1}' | xargs -P4 -I@ bash -c 'machine=@; svc=${machine//\//-}; svc=${svc%:}; juju ssh ${machine%:} "sudo service jujud-machine-${svc}" restart'
```

Comment:
re: `grep -v 'Machine'` - I kinda like that the command does not process the "table header", even if it's filtered out by the awk condition. Maybe we can make it more clear that we want to skip first line by using something like `tail -n +2` or `sed 1 d`

re: `svc=${svc%:} - Yeah, this was part of the original command and I didnt feel confident enough to remove it.I thought that maybe it covers some corner case.

Reply:
Neither of the suggestions is a blocker for me. It makes sense to me to run as few processes as possible (ie. only `awk` instead of `grep+awk` or `tail/sed + awk`, etc.) but I'm fine. Same for the `svc` value. Knowing the possible values, I know they won't end with a colon.

Reply:
* Removed the `svc=${svc%:}.
* Replaced the `grep -v 'Machine' with `sed "1 d"`. Imo we don't need to optimize this for performance and it's cleaner this way.